0009/README.md

```diff
@@ -1,89 +1,346 @@
+    "fields":[
+        {
+            "name":"correlationId",
+            "type":"string",
+            "doc":"Message ID primarily used to track message processing across services"
+        },
+        {
+            "name":"timestamp",
+            "type":"long",
+            "doc":"Message creation UNIX timestamp in milliseconds"
+        },
+        {
+            "name":"timeout",
+            "type":"long",
+            "default":0,
+            "doc":"Amount of milliseconds since the timestamp until the message expires. Value of 0 is reserved to indicate no expiration."
+        },
+        {
+            "name":"appVersionName",
+            "type":"string",
+            "doc":"Application version name the endpoint connected with"
+        },
+        {
+            "name":"endpointId",
+            "type":[
+                "string",
+                "null"
+            ],
+            "doc":"Identifier of the endpoint that connected to the server"
+        },
+        {
+            "name":"originatorReplicaId",
+            "type":"string",
+            "doc":"Identifier of the service replica that generated the event"
+        }
+    ]
+}
+```
 
-NATS subject format:
 
-`kaa.v1.events.{originator-service-instance-name}.endpoint.lifecycle.registered`
+### Endpoint disconnected
 
-The Avro schema for the NATS message payload can be found [here](./registered.avsc).
+Endpoint disconnected event is published when endpoint communication session with server becomes unavailable because of a disconnect, timeout, or other reasons.
+The `{event-type}` is `disconnected`.
 
-### Endpoint updated application version
+Originators MUST use the following NATS subject format for endpoint disconnected events:
+```
+kaa.v1.events.{originator-service-instance-name}.endpoint.connectivity.disconnected
+```
+
+The NATS message payload is an Avro object with the following schema ([ep-disconnected.avsc](./ep-disconnected.avsc)):
+
+```json
+{
+    "namespace":"org.kaaproject.ipc.event.gen.v1.endpoint.connectivity",
+    "name":"DisconnectedEvent",
+    "type":"record",
+    "doc":"Endpoint(s) disconnected event message",
+    "fields":[
+        {
+            "name":"correlationId",
+            "type":"string",
+            "doc":"Message ID primarily used to track message processing across services"
+        },
+        {
+            "name":"timestamp",
+            "type":"long",
+            "doc":"Message creation UNIX timestamp in milliseconds"
+        },
+        {
+            "name":"timeout",
+            "type":"long",
+            "default":0,
+            "doc":"Amount of milliseconds since the timestamp until the message expires. Value of 0 is reserved to indicate no expiration."
+        },
+        {
+            "name":"endpointIds",
+            "doc":"Identifiers of the endpoint(s) that got disconnected from the server",
+            "type":{
+                "type":"array",
+                "items":"string",
```

Comment:
Is there a reason to make connected/disconnected events symmetrical? i.g., make both carry an array of endpointIds

Reply:
Normally you have endpoints connecting one by one. However, they can get disconnected together with a client session. Therefore the difference.

Reply:
I understand.

I think making `enpointIds` an array in `ConnectedEvent` would be more beautiful. First, that will make a symmetry in these events; second, as a bonus point that will allow us handling connection of multiple endpoints simultaneously (which might be useful for some future protocols)

Reply:
It's gonna be an array of EP ID - appVersionName tuples.